### ISSUE TITLE
NullPointerException when changing models and/or closing a model.

### ISSUE THREAD

**Issue by j26151** [ZeligsoftDev/CX4CBDDS#390]
# **Issue and tracking information**
I don't have a reliable reproducer for this issue. It is one that we see frequently but random. I have attached the stack trace.
[ErrorLog.txt](https://github.com/ZeligsoftDev/CX4CBDDS/files/8464525/ErrorLog.txt)

### Developer's Actual time spent on fix (hours)  

# **Issue reporter to provide a detailed description of the issue in the space below**

**Comment by eposse** [comment]
Hi @j26151 When you say "changing models" do you mean that you have two models open, are working on one, and then switch to the other by, for example, click on a tab with a diagram of the other model?

**Comment by j26151** [comment]
Yes, clicking on different models (different *.di files) that are already opened.  This is an error that I have definitely seen on papyrus  4.4.0 (2019061217). I cannot remember if I saw it in the newer version that the head of the papyrus branch is pointing to. I wish I had extra details to give you. 
To give a little more context on this issue. It is pretty common for our developers to be going about there work and to sporadically have null pointer exceptions occur (with the error dialog popping up). At best, this is just an annoyance, but at worst is makes our product look buggy and less refined. We see these when doing seemingly harmless things like closing models or changing models. We really wish there was a way to reduce the frequency of these errors but sometimes it is really hard to track down what causes them. We understand that it becomes just as hard for your team to fix these issues without reliable reproducers, but we are hoping that just by bringing this conversation up, we might be able to get some progress made.

**Comment by eposse** [comment]
I understand that the popup is annoying, and I agree that these popups shouldn't happen. Although in some cases, we don't control when these pop-ups are displayed.

I'm testing to see if I can reproduce it, but the trace doesn't show any of our code being involved, so it looks like it is a Papyrus issue, which means that any solution would require a patch to Papyrus.

It looks like it is related to [Bug 559862](https://bugs.eclipse.org/bugs/show_bug.cgi?id=559862) in Papyrus.

What's more, the exception is rather strange, as the code in question is the following in the `org.eclipse.papyrus.views.modelexplorer.ModelExplorerView.deactivate` method:

```java
		if (editingDomain != null) {
			editingDomain.removeResourceSetListener(resourceSetListener);
```

The NPE is in the second line, immediately after checking that editDomain is not null. So it would appear that this could happen only if there are multiple calls to the `deactivate` method in different threads on the same `ModelExplorerView` instance and there is a context switch between those two lines.

I'll see what else I can find.

**Comment by j26151** [comment]
Yes, we too think this is a problem in papyrus itself. We are certainly not placing the blame for this NPE on CX or Zeligsoft. We were hoping though that you may be able to find a fix in papyrus and provide a patch to us like you have in the past (I think?)

**Comment by eposse** [comment]
I'd like to reproduce it as close as you, so I'm trying to build the RPM package, but I forgot how you do it, can you remind me?

I have rpmbuild installed and tried

```bash
cd releng/com.zeligsoft.dds4ccm.update.axcioma/rpm_build/
rpmbuild -ba zeligsoftCX_axcioma.spec
```

but I get some errors about non-existing folders. Do I need to run rpmbuild from a different folder? Do I need to set some environment variables? Does papyrus need to be pre-installed at /opt/Papyrus?

Thanks.

**Comment by J17359** [comment]
@eposse Looking at your GitHub actions [yaml file](https://github.com/ZeligsoftDev/CX4CBDDS/actions/runs/2184580889/workflow), it looks like you [build the CX plugins](https://github.com/ZeligsoftDev/CX4CBDDS/actions/runs/2184580889/workflow#L62-L74) with maven and then [run rpmbuild](https://github.com/ZeligsoftDev/CX4CBDDS/actions/runs/2184580889/workflow#L100-L105) on the artifact location. See the links in my text above to see the exact code for those commands.

Internally, we do the same sort of thing but with our own version of your spec file. Our spec file actually calls all the maven commands and runs the build so that everything is encompassed in the spec file and we can simply "rpmbuild" it.

**Comment by eposse** [comment]
It looks like both @ysroh and I can reproduce a very similar issue. There is one small difference however: The exception trace you attached, it begins with 

```java
java.lang.NullPointerException
	at org.eclipse.papyrus.views.modelexplorer.ModelExplorerView.deactivate(ModelExplorerView.java:1005)
	at org.eclipse.papyrus.views.modelexplorer.ModelExplorerView.dispose(ModelExplorerView.java:1061)
        ...
```

but the one we are getting begins with

```java
java.lang.NullPointerException
	at org.eclipse.papyrus.infra.ui.lifecycleevents.SaveAndDirtyService.removeInputChangedListener(SaveAndDirtyService.java:488)
	at org.eclipse.papyrus.views.modelexplorer.ModelExplorerView.deactivate(ModelExplorerView.java:1005)
	at org.eclipse.papyrus.views.modelexplorer.ModelExplorerView.dispose(ModelExplorerView.java:1061)
        ...
```

So both are related and both occur when the ModelExplorer view of a model is being closed, but as you can see, the exact origin is different. If you get this again, can you double-check if the exception looks more like the first one or like the second one?

Thanks.

**Comment by j26151** [comment]
Sure, I will try to pass along more info if and when I see it. Which papyrus version did both of you see your exception on?

**Comment by eposse** [comment]
Thanks. We are on Papyrus 4.8 (release 2020-06).

**Comment by eposse** [comment]
Just to give a status update.

While I'm able to "reproduce" this issue, more precisely the one that produces the NPE from the SaveAndDirtyService, I don't know the exact steps. However, I've been able to establish that:

1. It happens when you have two models and one references the other
2. You edit the referenced model, either in its own editor or from the model explorer in the other model
3. You switch between the two models
4. The switch triggers a reload of both models
5. At some point, the SaveAndDirtyService (and possibly all services) associated to the referring model are disposed *before* the reload, which leads to the NPE

However I do not know why the services are being disposed, and it doesn't seem predictable.

I'll keep investigating.

**Comment by eposse** [comment]
Hi @j26151 and @J17359 . An update on this issue, with a couple of questions for you.

The problem arises from some race condition in Papyrus, but I am still unable to pinpoint the precise reason for this, which means that it's not clear what exactly to patch in Papyrus, but we have been discussing and it would seem that the main problem for you here is the annoying pop-up, so we thought about an alternative solution, which is to detect this exceptions and prevent the tool from launching the pop-up even though we wouldn't prevent the exception itself, and rather just log it. So my first question is if this approach is acceptable to you?

If it is acceptable, then we can override the error handling facility to filter the undesired exception(s), but there's a catch: the only way to do this requires us to define a proper Eclipse product (with a .product definition file), rather than only an update site with features. In case you are not familiar with these, here's a link that explains what they are: [What us an Eclipse product](https://wiki.eclipse.org/FAQ_What_is_an_Eclipse_product%3F). This may have some consequences: first, we would need to make some adjustments to our build, adding some POM file configuration, second, it could enable us to add branding to the tool (like a custom splash image, icons, etc.) I have not yet looked into what would be entailed for the RPM build, but I'm confident it wouldn't be a big issue. So we are considering making the tool into an Eclipse product. We have done it in the past for other Papyrus-based tools, so we have had experience with it. We would start by trying to build the tool this way on a separate branch. My second question is then if this is something that you would be ok with?

**Comment by j26151** [comment]
@eposse This solution sounds reasonable to me, however I definitely want @J17359 to think about it a bit before we fully commit. He is out on vacation at the moment and wont be back for another week or so.

Some comments/questions that I have:

1. Would this allow you to just catch this specific instance of the exception and not say, all null pointer exceptions? We definitely want real, important exceptions to make it to the user.
2. My biggest concern would be ensure that packaging and delivering Papyrus with this patch wouldn't become extraordinarily more complex or difficult to package or install.
3. A assume if we come across other exceptions like this we would very easily be able to hide them in the future if we have this eclipse product work already done.

**Comment by eposse** [comment]
Hi @j26151, my answers are below:

1. The idea is to make the system as customizable as possible. The implementation allows you to select the "undesirable" exceptions you want to filter and ignore (i.e. avoid the pop-up) while still logging them, and all other exceptions are delegated to the standard error handler. My current implementation allows you to programatically register these "undesirable exceptions" providing a) the exception class to filter, b) the fully qualified name of the method where the exception might arise and c) the maximum depth to look for such method in the exception trace. So for example, we can declare an exception cause in some method, in Eclipse, in Papyrus or any custom plugin, and it doesn't have to be the immediate cause of the exception (although it would be when the maximum depth selected is 1). Furthermore, it goes through the exception "caused by" chain, so it can also filter exceptions that had been caught and then another exception was raised in its place. 
2. I think it should become easier to package and install, or at least not any more difficult than the current approach. By defining a product, we could generate a full RCP, that includes Eclipse, Papyrus, CX and dependencies and package it as a zip file.
3. Currently you register these "undesirable exceptions" with a method, but I'm thinking about declaring an extension point so that for example you can register your own undesirable exceptions in your own plugins. 

How does this sound to you?

**Comment by j26151** [comment]
@eposse I think all of this sounds great. Again, I would like Michael to think this over when he returns on Wednesday before we move forward.

**Comment by eposse** [comment]
I've opened PR #402 with the new branch.

It doesn't yet include the custom error handler to deal with this issue, but you can check how the all-in-one RCP product is built and test it. See the pull request for a description of what it includes and where to find the generated zip file.

**Comment by J17359** [comment]
@eposse My goal is to test that new artifact this afternoon or tomorrow morning. I'll report back here after. Sorry for taking a few days to get around to this!

**Comment by J17359** [comment]
@eposse I'm continuing to walk through some of our normal development processes and so far everything looks good. I've built our SNA plugins inside your new product in both EL7 and EL8 successfully. I'm now using those built plugins to launch a second Eclipse instance to test the plugins that were built. I'll keep you posted how that goes.

As for the name of this new product, I'd prefer `papyrus-cx-axcioma` or `papyrus4cx-axcioma` or something where `papyrus` and `cx` are separated by some character for readability. But I don't have a strong preference one way or the other. We could probably drop `axcioma` from the name entirely since we don't support Papyrus with the ATCD version of CX internally. That ATCD version of CX is staying on RSA so I don't think we need that axcioma specifier any longer if we're lumping papyrus and CX into the same product. But that's up to you!

The only thing I noticed so far was a broken "Tasks" view in the C/C++ perspective. That's not critical since I don't think our users use that but it used to work in Papyrus when it was its own product. Just wanted to point that out.
![image](https://user-images.githubusercontent.com/46937599/170018806-5843d3c2-241d-4525-9665-5ab23fc5047e.png)

**Comment by J17359** [comment]
@eposse Ok I think I'm happy with the current state of things (aside from the broken Tasks view in the C/C++ perspective - which really isn't a huge deal but something we'd like fixed or hidden). I ran through a few normal developer scenarios and was able to perform them all without issue. It's not a perfect test by any means and I didn't do any actual modeling during my testing, but we will test that during our normal development cycle once this is in place.

I'll also do some more testing once the RPM is available since that's how we install CX 99% of the time and the only way our users would interact with your software.

**Comment by eposse** [comment]
@J17359 Thanks for looking into it. I'll change the name, and look into the problem with the tasks view. If you get a change to share the message that appears when you click on "Details..." in that view, as well as the error log, it would be great.

As for the custom error handler, which is why we are going this path, I have a preliminary version, but it is not yet ready. The idea is that you can register "undesirable exceptions" via an extension point (so that you can add your own, addressing a concern raised by @j26151). That part is not working yet.

I'll also update the RPM spec so you can test that.

**Comment by J17359** [comment]
@eposse I think this might be coming from our default_preferences.epf file we import into new workspaces automatically for our users. I think there's some saved mylyn preferences in there. I'll try and remove those references to see if that fixes our bug. It only happens with the default_preferences.epf file imported so I think there's nothing to do on your end for that. Sorry for assuming it was your bug!

**Comment by eposse** [comment]
@J17359 @j26151 I've added an RPM spec for the product and pushed it to PR #402. Our automated build on Github doesn't use it yet, but you can test it by first building with mvnw as usual and then rpmbuild on the new spec which is located under `releng/com.zeligsoft.papyrus-cx-axcioma.rcp.product/rpm_build`.

I used the following script to build the rpm:

```bash
#!/bin/bash
CX4CBDDSROOT=$HOME/git/CX4CBDDS
RPM_TAG=test
PROJECTDIR=$CX4CBDDSROOT/releng/com.zeligsoft.papyrus-cx.axcioma.rcp.product
SPEC=$PROJECTDIR/rpm_build/papyrus-cx-axcioma.spec

rpmbuild -ba --clean -vv --define "_projectdir $PROJECTDIR" --define "_git_tag $RPM_TAG" $SPEC
```

You can adjust the directories according to your environment.

One thing I noticed is that the build is rather slow in the "brp-java-repack-jars" phase. I've been looking into whether this could be fixed by disabling this phase altogether. 

The RPM is installed in `/opt/papyrus-cx-axcioma` and the executable is `papyrus-cx`.

**Comment by J17359** [comment]
> One thing I noticed is that the build is rather slow in the "brp-java-repack-jars" phase. I've been looking into whether this could be fixed by disabling this phase altogether.
@eposse you can disable this phase using `%define __jar_repack %{nil}` at the top of your spec file. We do that in some of our Java-based RPMs to speed things up. We haven't noticed any down-side to disabling that repack but I'm also not entirely sure what it does in the first place so maybe we just got lucky? Your mileage may vary!

I'm going to build your branch and generate the RPM this morning to verify things work as expected in addition to taking a look at your spec changes. I'll keep you posted on my test results!

**Comment by J17359** [comment]
@eposse Now that you're packaging Papyrus with CX, could you add a Papyrus desktop file so we can see it in the Activities bar of the OS? For example, we had the following in our RPM spec that packages papyrus that I imagine could be modified for your specific use:
```
%{__mkdir_p} %{buildroot}/usr/share/applications/
echo "[Desktop Entry]
Encoding=UTF-8
Name=Papyrus IDE
Comment=Papyrus IDE
Exec=/opt/Papyrus/papyrus
Icon=/opt/Papyrus/icon.xpm
Categories=Application;Development;Java;IDE
Version=1.0
Type=Application
Terminal=false
StartupNotify=true" > %{buildroot}/usr/share/applications/papyrus.desktop
```

**Comment by eposse** [comment]
@J17359 I've added the desktop file and the define to skip the jar repack phase. It works nicely. Let me know if this (PR #402) is ok to merge now. Note that this still doesn't include PR #403 which works but Paul has asked for a couple of changes.

**Comment by J17359** [comment]
> @J17359 I've added the desktop file and the define to skip the jar repack phase. It works nicely. Let me know if this (PR #402) is ok to merge now. Note that this still doesn't include PR #403 which works but Paul has asked for a couple of changes.

I think PR #402 is ok to merge assuming your team has reviewed. Once it's on the papyrus branch, I'll generate the RPM on our end and start to wrap our plugins into that new product instead of papyrus and will let you know how that goes.

**Comment by eposse** [comment]
@J17359 @j26151 Both Paul and Young-Soo have approved merging PR #402 (the product build) so I've merged it.

I also rebased #403 and it is ready for you to test it.

Here's what I did to test it:
1. Created a model A with a single struct CXStruct1 with a single field CXField1, and gave it a Dynamic pathmap
2. Created a model B, imported model A, added a CXArray whose members are of type A::CXStruct1 (the type from model A)
3. Saved both
4. Edited the type of A::CXStruct1::CXField1 from the model explorer in model B, and saved
5. Switched between models several times
6. Edited the type of A::CXStruct1::CXField1 from the model explorer in model A, and saved
7. Switched between models several times
8. Repeated steps 4 to 7 until the model explorer of one of the two models became blank and closed that model

At this point, the "old" version would pop-up a dialog with an NPE.
With the current fix in #403 it will not pop-up anything but it will log the exception in the Error log.

**Comment by eposse** [comment]
(In reply to comment #32)

No problem. I'll be working on #397 in the meantime.

**Comment by J17359** [comment]
@eposse Would you guys be willing to install an XML Editor for us in your new Papyrus-CX product? We used to do it ourselves but are missing a dependency in the new Papyrus (it used to come with the old Papyrus). Rather than having you install the dependency, I wonder if it would be easier to simply install the XML Editor plugin for us?

The update site for the plugin: https://download.eclipse.org/webtools/repository/latest/
The feature group to install from update site: `org.eclipse.wst.xml_ui.feature.feature.group`
![image](https://user-images.githubusercontent.com/46937599/172372726-76635a59-80b1-4f3e-aec3-74768c67c80c.png)

We're still rather new to maven but I imagine it's easier for you all to pre-package that plugin during your RCP build rather than us maintaining a zipped update site that needs dependencies pre-installed since we're in a sandboxed environment sometimes and can't pull those from the web.

**Comment by eposse** [comment]
@J17359 Yes, no problem. It's quite easy. I'll have a PR for you shortly. If there are any other tools you'd like installed and prepackaged just let us know.

**Comment by J17359** [comment]
@eposse - Could you remind me what PRs are left for this ticket? I know we've had a few go in for the Papyrus product reorg but we don't have anything done for the NPE right? Just making sure there's nothing more for us to test at the moment!

**Comment by eposse** [comment]
@J17359 Yes we have fixed the NPE issue. See pull request #403. Also see the comment [above](https://github.com/ZeligsoftDev/CX4CBDDS/issues/390#issuecomment-1146410387) where I described how I tested it. I believe you were going to test it on your side. Let me know if it works for you.

**Comment by J17359** [comment]
> @J17359 Yes we have fixed the NPE issue. See pull request #403. Also see the comment [above](https://github.com/ZeligsoftDev/CX4CBDDS/issues/390#issuecomment-1146410387) where I described how I tested it. I believe you were going to test it on your side. Let me know if it works for you.

Have you rebased #403 since the last PR with the xml-plugin addition? I can generate the RPM from that PR today and get Eric or myself to test today or tomorrow

**Comment by eposse** [comment]
@J17359 I've merged the PR with the XML feature and rebased #403. The build is done and the RPM can be downloaded here: [https://github.com/ZeligsoftDev/CX4CBDDS/suites/6929687783/artifacts/269689309](https://github.com/ZeligsoftDev/CX4CBDDS/suites/6929687783/artifacts/269689309)

**Comment by J17359** [comment]
Great, thanks! Copying that artifact into our environment now. I hope to have the NPE fix tested by EOD tomorrow. I'll keep you posted.

**Comment by j26151** [comment]
So, when I attempted to follow your instructions to recreate with the RPM that you linked above 2 notable things happened:

1. Before peforming step 5, I saved the model after step 4. This caused a null pointer exception that showed a pop up. 
![image](https://user-images.githubusercontent.com/49879646/173661864-eed09cb9-1b69-478e-8089-6f1faac5b810.png)


2. I ignored this, did step 4 again. Saved. No issue. But when I switched models I saw that the model explorer went blank. When I closed the model, I received another null pointer exception:
 ![image](https://user-images.githubusercontent.com/49879646/173661733-7f384675-443c-48d3-a3f9-f39f5d47b262.png)

**Comment by j26151** [comment]
One concern I have with these errors specifically is that this is all related to attempting to modify a model through a dynamic pathmap relationship. I thought we had this implemented by @ysroh a few months ago. Why are we getting exceptions doing these fairly routine operations?

**Comment by j26151** [comment]
Yes, I received the pop-up with the second one. This is the situation that I would have expected to not get a pop up, but I did get one. I wonder if the pop up occured becuase there was still a previously un-removed exception from my first point? I am not sure. But I did get a pop up.

**Comment by eposse** [comment]
I doubt it's related to the first, but now I am experiencing an issue with the extension point to register exceptions which I didn't have before and which would explain why the filtering is not working so I'm investigating. I'll post back when I figure it out.

**Comment by j26151** [comment]
Sorry, so it looks like I may have received 2 exceptions at the same time for the second case. The one that you guys have specifically hidden was not shown. But I got the following exception as well:
![image](https://user-images.githubusercontent.com/49879646/173666801-7cde1534-cce2-4c53-afba-e43459b60765.png)

So it looks like your filter worked on the one exception, but it is possible to get two exceptions for that issue, where the second one shows.

**Comment by eposse** [comment]
The ServiceNotFound exceptions are very common in Papyrus and I've never seen one causing a pop-up.

But I think I found the reason why the pop-up may have occurred. It seems like it was not packaged in the RCP. I'm adding now the decorator exception and rebuilding.

**Comment by eposse** [comment]
Ok, I can confirm now that the new error handler was missing from the packaging in the previous build, which would explain the pop-ups.

I'm now pushing the fix and the updated package would be available in about 20~25min after the build.

**Comment by eposse** [comment]
The build is done. You can get the latest RPM [here](https://github.com/ZeligsoftDev/CX4CBDDS/suites/6932622817/artifacts/269870874).

**Comment by j26151** [comment]
This looks good to me. I verified that the error comes up in the Error Log but there is no pop up.

**Comment by J17359** [comment]
@eposse I think we're cool for you to merge https://github.com/ZeligsoftDev/CX4CBDDS/pull/403 when you're ready. Eric tested that RPM yesterday.

**Comment by eposse** [comment]
I've merged it. So we can close this issue?

If there are new pop-ups, let us know.

**Comment by j26151** [comment]
Yes, you can close this issue.
